naga/src/back/hlsl/help.rs

```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
+                        writeln!(self.out, "{level}return asint(~asuint(val) + 1u);",)?;
+                        writeln!(self.out, "}}")?;
+                        writeln!(self.out)?;
+                    }
+                    _ => {}
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(super) fn write_wrapped_binary_ops(
+        &mut self,
+        module: &crate::Module,
+        func_ctx: &FunctionCtx,
+    ) -> BackendResult {
+        for (expr_handle, expression) in func_ctx.expressions.iter() {
+            if let crate::Expression::Binary { op, left, right } = *expression {
+                let expr_ty = func_ctx.resolve_type(expr_handle, &module.types);
+                let left_ty = func_ctx.resolve_type(left, &module.types);
+                let right_ty = func_ctx.resolve_type(right, &module.types);
+
+                match (
+                    op,
+                    func_ctx
+                        .resolve_type(expr_handle, &module.types)
+                        .scalar_kind(),
+                ) {
+                    (
+                        crate::BinaryOperator::Divide,
+                        Some(kind @ crate::ScalarKind::Sint | kind @ crate::ScalarKind::Uint),
+                    ) => {
+                        let left_wrapped_ty = match *func_ctx.resolve_type(left, &module.types) {
+                            crate::TypeInner::Scalar(scalar) => (scalar, None),
+                            crate::TypeInner::Vector { size, scalar } => (scalar, Some(size)),
+                            _ => unreachable!(),
+                        };
+                        let right_wrapped_ty = match *func_ctx.resolve_type(right, &module.types) {
+                            crate::TypeInner::Scalar(scalar) => (scalar, None),
+                            crate::TypeInner::Vector { size, scalar } => (scalar, Some(size)),
+                            _ => unreachable!(),
+                        };
+                        let wrapped = WrappedBinaryOp {
+                            op,
+                            left_ty: left_wrapped_ty,
+                            right_ty: right_wrapped_ty,
+                        };
+                        if !self.wrapped.binary_op.insert(wrapped) {
+                            continue;
+                        }
+
+                        self.write_value_type(module, expr_ty)?;
+                        write!(self.out, " {DIV_FUNCTION}(")?;
+                        self.write_value_type(module, left_ty)?;
+                        write!(self.out, " lhs, ")?;
+                        self.write_value_type(module, right_ty)?;
+                        writeln!(self.out, " rhs) {{")?;
+                        let level = crate::back::Level(1);
+                        match kind {
+                            crate::ScalarKind::Sint => {
+                                let min = 2u64.pow(expr_ty.scalar_width().unwrap() as u32 * 8 - 1);
```

Comment:
The `unwrap` is kind of a code smell: if you know it's not going to panic, then you must have done some prior match that you ought to have been able to get the data out of.

This is actually the negative of the minimum, so the name is a little confusing.

Suggestion: drop the explicit `-` in the format string and then use:

```suggestion
                                let min = -1_i64 << (expr_ty.scalar_width().unwrap() as u32 * 8 - 1);
```

That should get you the right minimum even for 64-bit integer types.